File: clients/ts/src/events.ts
/**
 * The session event stream: SSE consumption over platform fetch streaming,
 * with the Go client's cursor integrity rules. The daemon's terminal
 * transport signals (oap-overflow, oap-replay-gap) are framing, not OAP
 * envelopes; they surface as typed errors carrying the reconnect cursor.
 */

import type { FetchResponse, OapClient, StreamReader } from './client.js';
import {
  AbortedError,
  DisconnectError,
  MalformedFrameError,
  OverflowError,
  ReplayGapError,
  ResumeMismatchError,
  DuplicateSequenceError,
  SequenceGapError,
  ServerError,
} from './errors.js';
import { SSEParser, type SSEFrame } from './sse.js';
import { EnvelopeType, parseEnvelope, type Envelope } from './protocol.js';
import type { OapSession } from './session.js';

/** SSE event names for the daemon's terminal transport signals. */
const SIGNAL_OVERFLOW = 'oap-overflow';
const SIGNAL_GAP = 'oap-replay-gap';

/** The first wait between reconnect attempts after a transport failure; it doubles up to the cap. */
const INITIAL_RECONNECT_BACKOFF_MS = 100;
const MAX_RECONNECT_BACKOFF_MS = 1_000;
/**
 * Bounds consecutive reconnects whose connection ends without delivering
 * anything: a cursor at a settled run's tail produces empty replays forever,
 * and an endless silent reconnect loop is worse than a reported disconnect.
 */
const MAX_EMPTY_CYCLES = 3;

/** events()/eventsAfter() options. */
export interface EventsOptions {
  /** Aborts the stream: iteration raises AbortedError and the connection is dropped. */
  signal?: AbortSignal;
}

/** One open SSE connection. */
interface Connection {
  readonly response: FetchResponse;
  readonly reader: StreamReader;
  readonly parser: SSEParser;
}

/**
 * EventStream is one ordered envelope stream over a session, consumed
 * exclusively through `for await`. Iteration ends cleanly (the loop simply
 * finishes) once a run's terminal event has been delivered; every other
 * error is terminal for the stream.
 */
export class EventStream implements AsyncIterable<Envelope> {
  /**
   * Resolves once the initial subscription is live — the daemon has
   * registered it — or rejects with the failure that prevented it. Await
   * this before submitting when the consuming loop does not start at once.
   */
  readonly ready: Promise<void>;

  private readonly session: OapSession;
  private readonly client: OapClient;
  private readonly signal?: AbortSignal;
  private readonly strict: boolean;
  private readonly startAfter: number | null;

  private conn: Connection | null = null;
  /** The last observed (run, sequence); run is empty until the first envelope fixes it. */
  private runId = '';
  private lastSeq = 0;
  /** Marks a connection opened with a cursor: its first envelope must continue runId at lastSeq+1 exactly. */
  private resumed = false;
  /** Distinguishes the initial connect (whose transport failures surface at once) from reconnects (which back off and retry). */
  private everConnected = false;
  /** Records that the replay-from-start reconnect fallback was already tried, so a session without a run parks live instead of retrying the speculative cursor forever. */
  private speculated = false;
  /** Records that a run's terminal envelope was delivered, so a subsequent end of stream is the documented clean end, not a drop. */
  private terminal = false;
  /** Counts envelopes delivered by the current connection. */
  private connEvents = 0;
  private emptyCycles = 0;
  private backoffMs = 0;
  private buffered: SSEFrame[] = [];
  private iterator: AsyncIterator<Envelope> | null = null;

  constructor(
    session: OapSession,
    client: OapClient,
    options: EventsOptions,
    cursor: { startAfter: number | null; runId: string },
  ) {
    this.session = session;
    this.client = client;
    this.signal = options.signal;
    this.strict = client.strictResume;
    this.startAfter = cursor.startAfter;
    this.runId = cursor.runId;
    // A manual cursor anchors the sequence expectation: the replay must
    // continue at startAfter + 1, not at 1.
    if (cursor.startAfter !== null) this.lastSeq = cursor.startAfter;
    const established = this.connect();
    this.ready = established.then(() => undefined);
    // Neither the consuming loop nor `ready` is awaited in every usage shape;
    // an unobserved initial-connect failure must not crash the process.
    this.ready.catch(() => {});
  }

  [Symbol.asyncIterator](): AsyncIterator<Envelope> {
    // One stream, one iterator: a second loop resumes the same position
    // instead of rewinding or double-reading the wire.
    if (!this.iterator) this.iterator = this.iterate();
    return this.iterator;
  }

  private async *iterate(): AsyncGenerator<Envelope, void, unknown> {
    try {
      // The initial connection was opened when events() was called; its
      // failure is surfaced here, by the first next().
      await this.ready;
      for (;;) {
        if (!this.conn) await this.connect();
        let envelope: Envelope;
        try {
          envelope = await this.poll();
        } catch (err) {
          if (!(err instanceof ConnectionDrop)) {
            // A daemon signal or a stream defect: surfaced, never retried.
            throw err;
          }
          this.closeConn();
          if (this.signal?.aborted) throw new AbortedError('event stream aborted');
          if (this.terminal) {
            // The stream's documented clean end at run terminality.
            return;
          }
          if (this.connEvents === 0) {
            this.emptyCycles += 1;
            if (this.emptyCycles >= MAX_EMPTY_CYCLES) {
              throw new DisconnectError(
                this.runId,
                this.lastSeq,
                new Error(`reconnected ${this.emptyCycles} times without receiving an event`),
              );
            }
          } else {
            this.emptyCycles = 0;
          }
          if (this.strict) {
            throw new DisconnectError(this.runId, this.lastSeq, err.cause);
          }
          // Invisible resume: reconnect with the cursor and continue.
          continue;
        }
        yield envelope;
      }
    } finally {
      this.closeConn();
    }
  }

  /**
   * Opens one SSE connection, retrying transport failures with backoff after
   * the first successful connection. A cursor is attached whenever the
   * stream holds one; a reconnect with no observed envelope yet replays the
   * current run from its start, falling back to a live subscription when
   * the session has no run to replay.
   */
  private async connect(): Promise<void> {
    for (;;) {
      this.throwIfAborted();
      const [after, speculative] = this.cursor();
      const path = this.session.path('/events') + (after === '' ? '' : `?after=${after}`);
      const headers: Record<string, string> = { Accept: 'text/event-stream' };
      if (after !== '') headers['Last-Event-ID'] = after;
      let response: FetchResponse;
      try {
        response = await this.client.request(path, { method: 'GET', headers, signal: this.signal });
      } catch (err) {
        if (this.signal?.aborted) throw new AbortedError('event stream aborted');
        if (!this.everConnected) throw err;
        await this.wait();
        continue;
      }
      if (response.status !== 200) {
        let body = '';
        try {
          body = await response.text();
        } catch {
          // The status itself already says the stream was refused.
        }
        const failure =
          this.client.failureError(response.status, body) ??
          new ServerError(response.status, '', `event stream returned status ${response.status}, want 200 OK`);
        if (speculative && failure.code === 'no_run_to_resume') {
          this.speculated = true;
          continue;
        }
        throw failure;
      }
      const contentType = response.headers.get('content-type') ?? '';
      const mediaType = contentType.split(';', 1)[0]?.trim().toLowerCase() ?? '';
      if (mediaType !== 'text/event-stream') {
        // A 200 with some other body would parse as an empty or garbage
        // stream — read errors that masquerade as drops, or a connection
        // that parks forever. The endpoint must declare the event stream.
        throw new ServerError(
          response.status,
          '',
          `event stream content type "${contentType}", want text/event-stream`,
        );
      }
      if (!response.body) {
        throw new ServerError(response.status, '', 'event stream response carries no body');
      }
      this.everConnected = true;
      this.conn = { response, reader: response.body.getReader(), parser: new SSEParser() };
      this.resumed = after !== '';
      this.connEvents = 0;
      this.backoffMs = 0;
      return;
    }
  }

  /**
   * Reports the reconnect cursor for the next connection. It is empty for a
   * fresh live subscription; a reconnect that has observed nothing
   * speculatively replays the current run from its start, so envelopes
   * emitted during the disconnect are not missed.
   */
  private cursor(): [string, boolean] {
    if (this.everConnected && this.runId !== '') return [String(this.lastSeq), false];
    if (this.startAfter !== null) return [String(this.startAfter), false];
    if (this.everConnected && !this.speculated) return ['0', true];
    return ['', false];
  }

  /** Sleeps one backoff step, doubling up to the cap. */
  private wait(): Promise<void> {
    if (this.backoffMs === 0) this.backoffMs = INITIAL_RECONNECT_BACKOFF_MS;
    else if (this.backoffMs < MAX_RECONNECT_BACKOFF_MS) this.backoffMs *= 2;
    const delay = this.backoffMs;
    return new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.signal?.removeEventListener('abort', onAbort);
        resolve();
      }, delay);
      const onAbort = () => {
        clearTimeout(timer);
        reject(new AbortedError('event stream aborted during reconnect'));
      };
      if (this.signal) {
        if (this.signal.aborted) {
          clearTimeout(timer);
          reject(new AbortedError('event stream aborted'));
          return;
        }
        this.signal.addEventListener('abort', onAbort, { once: true });
      }
    });
  }

  private closeConn(): void {
    if (this.conn) {
      // Release the connection; a released body read no longer delivers.
      void this.conn.reader.cancel?.(undefined).catch(() => {});
      this.conn = null;
      this.buffered = [];
    }
  }

  private throwIfAborted(): void {
    if (this.signal?.aborted) throw new AbortedError('event stream aborted');
  }

  /**
   * Reads frames until one envelope is delivered, a terminal signal or
   * stream defect is found, or the connection ends.
   */
  private async poll(): Promise<Envelope> {
    for (;;) {
      while (this.buffered.length > 0) {
        const frame = this.buffered.shift();
        if (frame === undefined) break;
        const envelope = this.handleFrame(frame);
        if (envelope) return envelope;
      }
      const conn = this.conn;
      if (!conn) throw new Error('client: event stream poll without a connection');
      let read: { done: boolean; value?: Uint8Array };
      try {
        read = await conn.reader.read();
      } catch (err) {
        throw new ConnectionDrop(err);
      }
      if (read.done) {
        // The document ended: a frame whose blank line lands at the very
        // end still dispatches, then the connection has ended.
        const tail = conn.parser.finish();
        if (tail.length > 0) {
          this.buffered.push(...tail);
          continue;
        }
        throw new ConnectionDrop(undefined);
      }
      if (read.value) this.buffered.push(...conn.parser.push(read.value));
    }
  }

  /** Applies one frame; returns the envelope it delivered, or undefined for a skipped frame. */
  private handleFrame(frame: SSEFrame): Envelope | undefined {
    switch (frame.event) {
      case SIGNAL_OVERFLOW: {
        const signal = decodeSignal<OverflowSignal>(frame);
        let runId = signal.run_id ?? '';
        let lastSequence = signal.last_sequence ?? 0;
        if (this.runId !== '') {
          // The hub names the run current at signal time, which can be a
          // newer run than the one this connection was consuming while
          // last_sequence counts the consumed run's envelopes. The stream's
          // own cursor is the pair it actually delivered, so recovery
          // resumes the run that overflowed.
          runId = this.runId;
          lastSequence = this.lastSeq;
        }
        throw new OverflowError(runId, lastSequence, signal.message ?? '');
      }
      case SIGNAL_GAP: {
        const signal = decodeSignal<GapSignal>(frame);
        throw new ReplayGapError(
          this.runId,
          signal.requested_after ?? 0,
          signal.oldest_available ?? 0,
          signal.latest_available ?? 0,
          signal.message ?? '',
        );
      }
      case 'message': {
        let parsed: Envelope;
        try {
          parsed = parseEnvelope(frame.data);
        } catch (err) {
          throw new MalformedFrameError('message frame is not an envelope', err);
        }
        this.deliver(parsed, frame);
        return parsed;
      }
      default:
        // An unknown named event is framing the client does not define;
        // skipping it keeps the stream forward-compatible.
        return undefined;
    }
  }

  /** Applies the stream's cursor integrity rules to one envelope and records its position. */
  private deliver(envelope: Envelope, frame: SSEFrame): void {
    // An envelope naming another session never belongs on this stream: a
    // misrouted stream must not deliver another session's content and
    // interaction requests as this one's.
    if (envelope.session_id !== this.session.id) {
      throw new MalformedFrameError(
        `envelope for session "${envelope.session_id ?? ''}" on the "${this.session.id}" stream`,
      );
    }
    // Every envelope on this stream is a sequenced run event; one without a
    // sequence cannot be positioned, and delivering it would leave the
    // cursor behind it — a later resume would replay it without any way to
    // detect the duplicate. Sequences start at one.
    const sequence = envelope.sequence;
    if (typeof sequence !== 'number' || !Number.isInteger(sequence) || sequence <= 0) {
      throw new MalformedFrameError('event envelope carries no sequence');
    }
    if (frame.hasId) {
      if (!/^\d+$/.test(frame.lastId)) {
        throw new MalformedFrameError(`frame id "${frame.lastId}" is not a sequence`);
      }
      const id = Number(frame.lastId);
      if (sequence !== id) {
        throw new MalformedFrameError(`frame id ${id} disagrees with envelope sequence ${sequence}`);
      }
    }
    let resumed = this.resumed;
    if (resumed) {
      this.resumed = false;
      if (this.runId !== '' && envelope.run_id !== this.runId) {
        throw new ResumeMismatchError(this.lastSeq, this.runId, envelope.run_id ?? '', sequence);
      }
      // A cursor-bearing connection — including one replaying after zero —
      // requests the run from a known position, unlike a fresh live
      // subscription that may join mid-run: its first envelope must continue
      // the cursor exactly, or envelopes were skipped.
      if (sequence !== this.lastSeq + 1) {
        throw new SequenceGapError(envelope.run_id ?? '', this.lastSeq + 1, sequence);
      }
    }
    if (envelope.run_id !== this.runId) {
      if (resumed && this.runId === '') {
        // The first envelope of a cursor-carrying stream names the run the
        // cursor belongs to: the run is adopted, and the sequence
        // expectation stays bound to the cursor.
        this.runId = envelope.run_id ?? '';
        this.terminal = isTerminal(envelope.type);
      } else if (this.runId === '') {
        // The stream's first observed envelope may join a run in progress:
        // sequences before the join were never deliverable to this
        // subscription, so the join position becomes the baseline.
        this.runId = envelope.run_id ?? '';
        this.lastSeq = 0;
        this.terminal = isTerminal(envelope.type);
      } else {
        // A live transition to a new run starts a fresh sequence space, and
        // this stream was attached throughout, so it must witness the run
        // from its first envelope: anything later means the opening events
        // were lost.
        if (sequence !== 1) {
          throw new SequenceGapError(envelope.run_id ?? '', 1, sequence);
        }
        this.runId = envelope.run_id ?? '';
        this.lastSeq = 0;
        this.terminal = isTerminal(envelope.type);
      }
    } else if (isTerminal(envelope.type)) {
      this.terminal = true;
    }
    // Run sequences are contiguous: within one run every envelope carries
    // the previous sequence plus one. A regression is a replay defect, and a
    // skip means an envelope was lost — advancing past it would hide it from
    // the consumer and from a later cursor resume, so both surface.
    if (this.lastSeq > 0) {
      if (sequence <= this.lastSeq) {
        throw new DuplicateSequenceError(envelope.run_id ?? '', sequence);
      }
      if (sequence !== this.lastSeq + 1) {
        throw new SequenceGapError(envelope.run_id ?? '', this.lastSeq + 1, sequence);
      }
    }
    this.lastSeq = sequence;
    this.connEvents += 1;
  }
}

interface OverflowSignal {
  run_id?: string;
  last_sequence?: number;
  message?: string;
}

interface GapSignal {
  requested_after?: number;
  oldest_available?: number;
  latest_available?: number;
  message?: string;
}

function decodeSignal<Signal extends OverflowSignal | GapSignal>(frame: SSEFrame): Signal {
  let value: unknown;
  try {
    value = JSON.parse(frame.data);
  } catch (err) {
    throw new MalformedFrameError('signal frame payload did not decode', err);
  }
  if (typeof value !== 'object' || value === null) {
    throw new MalformedFrameError('signal frame payload did not decode');
  }
  return value as Signal;
}

/** Wraps the read error (clean end included) that ended one SSE connection: the stream may resume from its cursor. */
class ConnectionDrop {
  constructor(readonly cause: unknown) {}
}

function isTerminal(type: string): boolean {
  return type === EnvelopeType.RunCompleted || type === EnvelopeType.RunFailed || type === EnvelopeType.RunCancelled;
}

File: client/events.go
package client

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"

	"github.com/lsm/open-agent-protocol/protocol"
)

// SSE event names for the daemon's terminal transport signals. They are
// framing, not OAP envelopes.
const (
	signalOverflow = "oap-overflow"
	signalGap      = "oap-replay-gap"
)

const (
	// initialReconnectBackoff is the first wait between reconnect attempts
	// after a transport failure; it doubles up to maxReconnectBackoff.
	initialReconnectBackoff = 100 * time.Millisecond
	maxReconnectBackoff     = time.Second
	// maxEmptyCycles bounds consecutive reconnects whose connection ends
	// without delivering anything: an EventsAfter cursor at a settled run's
	// tail produces empty replays forever, and an endless silent reconnect
	// loop is worse than a reported disconnect.
	maxEmptyCycles = 3
)

// Events returns the session's event stream as a live subscription. The
// subscription is opened before Events returns — the daemon registers it once
// the response begins — so the canonical order (Events before Submit) cannot
// miss the run's first envelope; a subscription opened mid-run receives only
// events from that point on.
//
// Envelopes are returned in order. By default a dropped connection is resumed
// invisibly: the client reconnects with the last observed sequence as the
// cursor, the daemon replays the suffix, and the stream continues without
// duplicates. WithStrictResume turns this off and reports the drop instead.
// Next returns io.EOF once the stream ends at a run's terminal event. A
// connection failure while opening the subscription is reported by the first
// Next call.
func (s *Session) Events(ctx context.Context) *EventStream {
	stream := &EventStream{session: s, ctx: ctx, strict: s.client.strict}
	stream.establish()
	return stream
}

// EventsAfter returns the event stream replayed from a cursor: the current
// run's envelopes after the given sequence first, then live events. It is the
// manual resume path for consumers holding a cursor from an OverflowError,
// ReplayGapError, or DisconnectError.
func (s *Session) EventsAfter(ctx context.Context, after uint64) *EventStream {
	stream := &EventStream{session: s, ctx: ctx, strict: s.client.strict, startAfter: &after, lastSeq: after}
	stream.establish()
	return stream
}

// establish opens the initial connection so the subscription is live before
// the caller proceeds; its failure is surfaced by the first Next call.
func (es *EventStream) establish() {
	if err := es.connect(); err != nil {
		es.finished, es.err = true, err
	}
}

// EventStream is one ordered envelope stream over a session. It is not safe
// for concurrent use: exactly one goroutine consumes it through Next.
type EventStream struct {
	session    *Session
	ctx        context.Context
	strict     bool
	startAfter *uint64

	response *http.Response
	reader   *bufio.Reader
	// cursor is the last observed (run, sequence); runID is empty until the
	// first envelope fixes it.
	runID   protocol.RunID
	lastSeq uint64
	// resumed marks a connection opened with a cursor: its first envelope
	// must continue runID at lastSeq+1 exactly.
	resumed bool
	// everConnected distinguishes the initial connect (whose transport
	// failures surface at once) from reconnects (which back off and retry).
	everConnected bool
	// speculated records that the replay-from-start reconnect fallback was
	// already tried, so a session without a run parks live instead of
	// retrying the speculative cursor forever.
	speculated bool
	// terminal records that a run's terminal envelope was delivered, so a
	// subsequent end of stream is the documented clean end, not a drop.
	terminal bool
	// connEvents counts envelopes delivered by the current connection.
	connEvents  int
	emptyCycles int
	backoff     time.Duration

	finished bool
	err      error
}

// Next returns the next envelope. It returns io.EOF after the stream ends
// cleanly at a run's terminal event, and any other error is terminal for the
// stream: once Next reports an error other than io.EOF, subsequent calls
// return the same error.
func (es *EventStream) Next() (protocol.Envelope, error) {
	if es.finished {
		if es.err != nil {
			return protocol.Envelope{}, es.err
		}
		return protocol.Envelope{}, io.EOF
	}
	for {
		if es.response == nil {
			if err := es.connect(); err != nil {
				return protocol.Envelope{}, es.stop(err)
			}
		}
		envelope, err := es.poll()
		if err == nil {
			return envelope, nil
		}
		var drop *connectionDrop
		if !errors.As(err, &drop) {
			// A daemon signal or a stream defect: surfaced, never retried.
			return protocol.Envelope{}, es.stop(err)
		}
		es.closeResponse()
		if err := es.ctx.Err(); err != nil {
			return protocol.Envelope{}, es.stop(err)
		}
		if es.terminal {
			// The stream's documented clean end at run terminality.
			return protocol.Envelope{}, es.stop(io.EOF)
		}
		if es.connEvents == 0 {
			es.emptyCycles++
			if es.emptyCycles >= maxEmptyCycles {
				return protocol.Envelope{}, es.stop(&DisconnectError{
					RunID: es.runID, LastSequence: es.lastSeq,
					Cause: fmt.Errorf("reconnected %d times without receiving an event", es.emptyCycles),
				})
			}
		} else {
			es.emptyCycles = 0
		}
		if es.strict {
			return protocol.Envelope{}, es.stop(&DisconnectError{
				RunID: es.runID, LastSequence: es.lastSeq, Cause: drop.cause,
			})
		}
		// Invisible resume: reconnect with the cursor and continue.
	}
}

// stop makes err permanent for the stream: io.EOF is the clean end.
func (es *EventStream) stop(err error) error {
	es.closeResponse()
	es.finished = true
	es.err = err
	return err
}

// connect opens one SSE connection, retrying transport failures with backoff
// after the first successful connection. A cursor is attached whenever the
// stream holds one; a reconnect with no observed envelope yet replays the
// current run from its start, falling back to a live subscription when the
// session has no run to replay.
func (es *EventStream) connect() error {
	for {
		if err := es.ctx.Err(); err != nil {
			return err
		}
		after, speculative := es.cursor()
		response, err := es.open(after)
		if err != nil {
			var serverErr *ServerError
			if speculative && errors.As(err, &serverErr) && serverErr.Code == "no_run_to_resume" {
				es.speculated = true
				continue
			}
			if errors.As(err, &serverErr) {
				return err
			}
			if !es.everConnected {
				return err
			}
			if err := es.wait(); err != nil {
				return err
			}
			continue
		}
		es.everConnected = true
		es.response = response
		es.reader = bufio.NewReader(response.Body)
		es.resumed = after != ""
		es.connEvents = 0
		es.backoff = 0
		return nil
	}
}

// cursor reports the reconnect cursor for the next connection. It is empty
// for a fresh live subscription; a reconnect that has observed nothing
// speculatively replays the current run from its start (speculative true), so
// envelopes emitted during the disconnect are not missed.
func (es *EventStream) cursor() (cursor string, speculative bool) {
	switch {
	case es.everConnected && es.runID != "":
		return strconv.FormatUint(es.lastSeq, 10), false
	case es.startAfter != nil:
		return strconv.FormatUint(*es.startAfter, 10), false
	case es.everConnected && !es.speculated:
		return "0", true
	default:
		return "", false
	}
}

// open performs one GET on the session's event stream with the cursor both as
// the Last-Event-ID header and as the explicit ?after= query parameter.
func (es *EventStream) open(after string) (*http.Response, error) {
	path := es.session.path("/events")
	if after != "" {
		path += "?after=" + after
	}
	request, err := http.NewRequestWithContext(es.ctx, http.MethodGet, es.session.client.url(path), nil)
	if err != nil {
		return nil, err
	}
	if after != "" {
		request.Header.Set("Last-Event-ID", after)
	}
	request.Header.Set("Accept", "text/event-stream")
	response, err := es.session.client.eventsHTTP.Do(request)
	if err != nil {
		return nil, fmt.Errorf("client: stream session %s events: %w", es.session.id, err)
	}
	if response.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(response.Body)
		response.Body.Close()
		return nil, statusError(response, body)
	}
	return response, nil
}

// wait sleeps one backoff step, doubling up to the cap.
func (es *EventStream) wait() error {
	if es.backoff == 0 {
		es.backoff = initialReconnectBackoff
	} else if es.backoff < maxReconnectBackoff {
		es.backoff *= 2
	}
	timer := time.NewTimer(es.backoff)
	defer timer.Stop()
	select {
	case <-es.ctx.Done():
		return es.ctx.Err()
	case <-timer.C:
		return nil
	}
}

func (es *EventStream) closeResponse() {
	if es.response != nil {
		_ = es.response.Body.Close()
		es.response, es.reader = nil, nil
	}
}

// poll reads frames until one envelope is delivered, a terminal signal or
// stream defect is found, or the connection ends. A connectionDrop wraps the
// causes that mean "the connection ended, decide whether to resume".
func (es *EventStream) poll() (protocol.Envelope, error) {
	var (
		envelope protocol.Envelope
		failure  error
	)
	scanErr := scanSSE(es.reader, func(f frame) bool {
		switch f.event {
		case signalOverflow:
			signal, err := decodeSignal[overflowSignal](f.data)
			if err != nil {
				failure = err
			} else {
				failure = &OverflowError{RunID: protocol.RunID(signal.RunID), LastSequence: signal.LastSequence, Message: signal.Message}
			}
			return false
		case signalGap:
			signal, err := decodeSignal[gapSignal](f.data)
			if err != nil {
				failure = err
			} else {
				failure = &ReplayGapError{RequestedAfter: signal.RequestedAfter, OldestAvailable: signal.OldestAvailable, LatestAvailable: signal.LatestAvailable, Message: signal.Message}
			}
			return false
		case "message":
			parsed, err := protocol.ParseEnvelope(f.data)
			if err != nil {
				failure = &MalformedFrameError{Detail: "message frame is not an envelope", Cause: err}
				return false
			}
			if err := es.session.client.checkEnvelope(parsed, f.data); err != nil {
				failure = err
				return false
			}
			if err := es.deliver(parsed, f); err != nil {
				failure = err
				return false
			}
			envelope = parsed
			return false
		default:
			// An unknown named event is framing the client does not define;
			// skipping it keeps the stream forward-compatible.
			return true
		}
	})
	if failure != nil {
		es.closeResponse()
		return protocol.Envelope{}, failure
	}
	if scanErr != nil {
		es.closeResponse()
		return protocol.Envelope{}, &connectionDrop{cause: scanErr}
	}
	return envelope, nil
}

// deliver applies the stream's cursor integrity rules to one envelope and
// records its position.
func (es *EventStream) deliver(envelope protocol.Envelope, f frame) error {
	if f.hasID {
		id, err := strconv.ParseUint(f.lastID, 10, 64)
		if err != nil {
			return &MalformedFrameError{Detail: fmt.Sprintf("frame id %q is not a sequence", f.lastID), Cause: err}
		}
		switch {
		case envelope.Sequence == nil:
			return &MalformedFrameError{Detail: fmt.Sprintf("frame id %d frames an envelope with no sequence", id)}
		case *envelope.Sequence != id:
			return &MalformedFrameError{Detail: fmt.Sprintf("frame id %d disagrees with envelope sequence %d", id, *envelope.Sequence)}
		}
	}
	if es.resumed {
		es.resumed = false
		if envelope.Sequence == nil {
			return &MalformedFrameError{Detail: "a resumed stream must deliver sequenced envelopes"}
		}
		if es.runID != "" && envelope.RunID != es.runID {
			return &ResumeMismatchError{AfterSequence: es.lastSeq, ExpectedRunID: es.runID, ObservedRunID: envelope.RunID, ObservedSequence: *envelope.Sequence}
		}
	}
	if envelope.RunID != es.runID {
		// Sequences are per-run: a new run starts a fresh sequence space.
		es.runID = envelope.RunID
		es.lastSeq = 0
		es.terminal = isTerminal(envelope.Type)
	} else if isTerminal(envelope.Type) {
		es.terminal = true
	}
	if envelope.Sequence != nil {
		// Run sequences are contiguous: within one run every envelope carries
		// the previous sequence plus one. A regression is a replay defect, and
		// a skip means an envelope was lost — advancing past it would hide it
		// from the consumer and from a later cursor resume, so both surface.
		if es.lastSeq > 0 {
			switch {
			case *envelope.Sequence <= es.lastSeq:
				return &DuplicateSequenceError{RunID: envelope.RunID, Sequence: *envelope.Sequence}
			case *envelope.Sequence != es.lastSeq+1:
				return &SequenceGapError{RunID: envelope.RunID, Expected: es.lastSeq + 1, Observed: *envelope.Sequence}
			}
		}
		es.lastSeq = *envelope.Sequence
	}
	es.connEvents++
	return nil
}

func isTerminal(typ protocol.EnvelopeType) bool {
	switch typ {
	case protocol.TypeRunCompleted, protocol.TypeRunFailed, protocol.TypeRunCancelled:
		return true
	}
	return false
}

func decodeSignal[Signal any](data []byte) (Signal, error) {
	var signal Signal
	if err := json.Unmarshal(data, &signal); err != nil {
		var zero Signal
		return zero, &MalformedFrameError{Detail: "signal frame payload did not decode", Cause: err}
	}
	return signal, nil
}

type overflowSignal struct {
	RunID        string `json:"run_id"`
	LastSequence uint64 `json:"last_sequence"`
	Message      string `json:"message"`
}

type gapSignal struct {
	RequestedAfter  uint64 `json:"requested_after"`
	OldestAvailable uint64 `json:"oldest_available"`
	LatestAvailable uint64 `json:"latest_available"`
	Message         string `json:"message"`
}

// connectionDrop wraps the read error (io.EOF included) that ended one SSE
// connection: the stream may resume from its cursor.
type connectionDrop struct{ cause error }

func (e *connectionDrop) Error() string {
	return fmt.Sprintf("client: event stream connection ended: %v", e.cause)
}

func (e *connectionDrop) Unwrap() error { return e.cause }

// OverflowError is the daemon's oap-overflow signal: this connection's bounded
// buffer fell behind. LastSequence is the last sequence delivered on the
// stream; resume with EventsAfter and a cursor after it.
type OverflowError struct {
	RunID        protocol.RunID
	LastSequence uint64
	Message      string
}

func (e *OverflowError) Error() string {
	return fmt.Sprintf("client: event stream overflowed behind sequence %d (run %s); resume with a cursor after it", e.LastSequence, e.RunID)
}

// ReplayGapError is the daemon's oap-replay-gap signal: the requested cursor
// is no longer retained. OldestAvailable and LatestAvailable bound what is;
// a consumer that accepts the loss resumes with a cursor at or after
// OldestAvailable - 1.
type ReplayGapError struct {
	RequestedAfter  uint64
	OldestAvailable uint64
	LatestAvailable uint64
	Message         string
}

func (e *ReplayGapError) Error() string {
	floor := uint64(0)
	if e.OldestAvailable > 1 {
		floor = e.OldestAvailable - 1
	}
	return fmt.Sprintf("client: replay cursor %d expired (retained %d through %d); resume at or after %d", e.RequestedAfter, e.OldestAvailable, e.LatestAvailable, floor)
}

// DisconnectError reports a dropped event stream that was not resumed:
// strict mode reports every drop, and auto-resume reports a stream that ends
// repeatedly without events. RunID and LastSequence are the stream's last
// observed position; resume with EventsAfter(LastSequence).
type DisconnectError struct {
	RunID        protocol.RunID
	LastSequence uint64
	Cause        error
}

func (e *DisconnectError) Error() string {
	return fmt.Sprintf("client: event stream disconnected after sequence %d (run %s): %v", e.LastSequence, e.RunID, e.Cause)
}

func (e *DisconnectError) Unwrap() error { return e.Cause }

// MalformedFrameError reports a stream frame the client cannot interpret: a
// message frame that is not an envelope, a signal payload that does not
// decode, or an id field disagreeing with the envelope it frames.
type MalformedFrameError struct {
	Detail string
	Cause  error
}

func (e *MalformedFrameError) Error() string {
	if e.Cause != nil {
		return fmt.Sprintf("client: malformed event stream frame: %s: %v", e.Detail, e.Cause)
	}
	return fmt.Sprintf("client: malformed event stream frame: %s", e.Detail)
}

func (e *MalformedFrameError) Unwrap() error { return e.Cause }

// DuplicateSequenceError reports an envelope whose (run, sequence) position
// was already delivered: a resumed stream replayed what the consumer already
// saw, which is a wire defect to surface, not skip.
type DuplicateSequenceError struct {
	RunID    protocol.RunID
	Sequence uint64
}

func (e *DuplicateSequenceError) Error() string {
	return fmt.Sprintf("client: duplicate sequence %d in run %s", e.Sequence, e.RunID)
}

// SequenceGapError reports an envelope that skipped one or more sequences in
// its run: an envelope was lost in transit or never published, and advancing
// the cursor past the hole would silently drop it.
type SequenceGapError struct {
	RunID    protocol.RunID
	Expected uint64
	Observed uint64
}

func (e *SequenceGapError) Error() string {
	return fmt.Sprintf("client: run %s skipped from sequence %d to %d", e.RunID, e.Expected, e.Observed)
}

// ResumeMismatchError reports a replayed suffix that does not continue the
// stream it was asked to: the run changed under the cursor, or the first
// replayed sequence is not the cursor plus one.
type ResumeMismatchError struct {
	AfterSequence    uint64
	ExpectedRunID    protocol.RunID
	ObservedRunID    protocol.RunID
	ObservedSequence uint64
}

func (e *ResumeMismatchError) Error() string {
	return fmt.Sprintf("client: replay after sequence %d continued run %s at sequence %d, want run %s at %d", e.AfterSequence, e.ObservedRunID, e.ObservedSequence, e.ExpectedRunID, e.AfterSequence+1)
}
